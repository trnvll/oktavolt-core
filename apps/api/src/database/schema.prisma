// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider      = "prisma-client-js"
  binaryTargets = ["native", "rhel-openssl-1.0.x"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

model User {
  id          Int      @id @default(autoincrement())
  email       String   @unique
  auth0Id     String   @unique
  username    String   @unique
  phoneNumber String?  @unique
  profile     Profile?
  createdAt   DateTime @default(now())
  Event       Event?   @relation(fields: [eventId], references: [id])
  eventId     Int?
}

model Profile {
  id          Int          @id @default(autoincrement())
  userId      Int?         @unique
  firstName   String
  lastName    String
  location    String?
  skills      Skill[]
  interests   Interest[]
  projects    Project[]
  user        User?        @relation(fields: [userId], references: [id])
  connections Connection[]
  createdAt   DateTime     @default(now())
}

model Skill {
  id        Int     @id @default(autoincrement())
  name      String
  profileId Int
  profile   Profile @relation(fields: [profileId], references: [id])
}

model Interest {
  id        Int     @id @default(autoincrement())
  name      String
  profileId Int
  profile   Profile @relation(fields: [profileId], references: [id])
}

model Project {
  id          Int     @id @default(autoincrement())
  title       String
  description String
  profileId   Int
  profile     Profile @relation(fields: [profileId], references: [id])
}

model Connection {
  id          Int      @id @default(autoincrement())
  connectedTo Int
  profileId   Int
  profile     Profile  @relation(fields: [profileId], references: [id])
  createdAt   DateTime @default(now())
}

model Event {
  id          Int      @id @default(autoincrement())
  title       String
  description String
  date        DateTime
  attendees   User[]
  createdAt   DateTime @default(now())
}

model Message {
  id        Int      @id @default(autoincrement())
  name      String
  email     String
  content   String
  createdAt DateTime @default(now())
}
