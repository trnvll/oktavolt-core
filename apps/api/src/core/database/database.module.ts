import { Module } from '@nestjs/common'
import { drizzle } from 'drizzle-orm/postgres-js'
import postgres from 'postgres'
import { schema } from 'database'
import { envConfig } from '@/config/env/env.config'

export const DATABASE_CONN = 'DATABASE_CONN'

@Module({
  providers: [
    {
      provide: DATABASE_CONN,
      useFactory: async () => {
        const connection = postgres(
          process.env.DATABASE_URL ?? envConfig.get('DATABASE_URL'),
        )
        return drizzle(connection, { schema })
      },
    },
  ],
  exports: [DATABASE_CONN],
})
export class DatabaseModule {}
