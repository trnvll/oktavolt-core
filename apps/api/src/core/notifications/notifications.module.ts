import { Module } from '@nestjs/common'
import { NotificationsService } from '@/core/notifications/services/notifications.service'

@Module({
  providers: [NotificationsService],
  exports: [NotificationsService],
})
export class NotificationsModule {}
