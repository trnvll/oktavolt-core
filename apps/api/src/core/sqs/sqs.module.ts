import { Module } from '@nestjs/common'
import { SqsService } from '@/core/sqs/sqs.service'

@Module({
  providers: [SqsService],
  exports: [SqsService],
})
export class SqsModule {}
