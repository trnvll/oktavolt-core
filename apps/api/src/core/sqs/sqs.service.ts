import { Injectable } from '@nestjs/common'
import { SQS } from 'aws-sdk'
import { ConfigService } from '@nestjs/config'
import { AwsConfig } from '@/config/aws.config'
import { BaseConfig } from '@/config/base.config'

@Injectable()
export class SqsService {
  private readonly sqs: SQS
  private readonly queueUrl: string

  constructor(private readonly configService: ConfigService) {
    const baseConfig = configService.get<BaseConfig>('base')
    const awsConfig = configService.get<AwsConfig>('aws')

    if (!awsConfig?.region) {
      throw new Error('AWS region is missing.')
    }

    if (!baseConfig?.sqsUserEventsQueueUrl) {
      throw new Error('SQS user events queue URL is missing.')
    }

    this.sqs = new SQS({ region: awsConfig.region })
    this.queueUrl = baseConfig.sqsUserEventsQueueUrl
  }

  async sendMessage(messageBody: any): Promise<void> {
    const params = {
      QueueUrl: this.queueUrl,
      MessageBody: JSON.stringify(messageBody),
    }
    await this.sqs.sendMessage(params).promise()
  }
}
