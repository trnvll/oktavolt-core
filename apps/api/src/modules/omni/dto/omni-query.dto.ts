import { IsString } from 'class-validator'

export class OmniQueryDto {
  @IsString()
  query: string
}
