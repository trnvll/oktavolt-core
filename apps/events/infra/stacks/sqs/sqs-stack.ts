import { Stack, StackProps, Duration } from 'aws-cdk-lib'
import { Construct } from 'constructs'
import * as sqs from 'aws-cdk-lib/aws-sqs'

export class EventsSQSStack extends Stack {
  public readonly queue: sqs.Queue

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props)

    this.queue = new sqs.Queue(this, 'EventsQueue', {
      visibilityTimeout: Duration.seconds(300),
    })
  }
}
