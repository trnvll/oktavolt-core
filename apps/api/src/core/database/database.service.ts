import { Inject, Injectable } from '@nestjs/common'
import { DATABASE_CONN } from '@/core/database/database.module'
import { schema } from 'database'
import { PostgresJsDatabase } from 'drizzle-orm/postgres-js'

@Injectable()
export class DatabaseService {
  constructor(
    @Inject(DATABASE_CONN) readonly db: PostgresJsDatabase<typeof schema>,
  ) {}
}
