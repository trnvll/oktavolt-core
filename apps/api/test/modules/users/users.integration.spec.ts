import { INestApplication } from '@nestjs/common'
import { CreateUsersDto } from '@/modules/users/dtos/create-user.dto'
import request from 'supertest'
import { faker } from '@faker-js/faker'
import { DatabaseService } from '@/core/database/database.service'
import { eq } from 'drizzle-orm'
import { Users } from 'database'
import { NotificationsService } from '@/core/notifications/services/notifications.service'
import { SqsService } from '@/core/sqs/sqs.service'
import { UserEmbeddingsService } from '@/modules/users/services/user-embeddings.service'
import { EventTypeEnum } from 'shared'
import { setupTestApp } from '../../app.helper'
import { describe, it, expect, beforeAll, afterAll } from 'vitest'

describe('UsersController (e2e)', () => {
  let app: INestApplication
  let context: {
    database: DatabaseService
    notificationService: NotificationsService
    sqsService: SqsService
    userEmbeddingsService: UserEmbeddingsService
    // userEventsQueue: Queue
  }
  let teardown: () => Promise<void>

  beforeAll(async () => {
    const testApp = await setupTestApp()
    app = testApp.app
    teardown = testApp.teardown

    context = {
      database: testApp.resolve(DatabaseService, true),
      notificationService: testApp.resolve(NotificationsService),
      sqsService: testApp.resolve(SqsService),
      userEmbeddingsService: testApp.resolve(UserEmbeddingsService),
      // userEventsQueue: app.get(QueueEnum.UserEvents),
    }
  })

  afterAll(async () => {
    await teardown()
  })

  it('should create multiple users and trigger the appropriate events', async () => {
    const createUsersDto: CreateUsersDto = {
      data: [
        {
          email: faker.internet.email(),
          firstName: faker.person.firstName(),
          lastName: faker.person.lastName(),
          phone: faker.phone.number(),
          dateOfBirth: faker.date.past(),
        },
      ],
    }

    const response = await request(app.getHttpServer())
      .post('/users')
      .send(createUsersDto)
      .expect(201)

    // Business rules:
    // 1. User data should be saved in the database.
    const savedUser = await context.database.db.query.users.findFirst({
      where: eq(Users.email, createUsersDto.data[0].email),
    })
    expect(savedUser).toBeDefined()
    if (!savedUser) throw new Error('User not found.')
    expect(savedUser.email).toBe(createUsersDto.data[0].email)

    // 2. Background job should have been triggered to send a welcome email to the user.
    // 3. Background job should have been triggered to generate and store user embeddings.
    // 4. Background job should have been triggered to store the user created event.
    // To verify these background jobs, we need to mock the respective services and check if the methods were called.

    const sendWelcomeEmailSpy = jest.spyOn(
      context.notificationService,
      'sendEmailNotification',
    )
    const generateEmbeddingsSpy = jest.spyOn(
      context.userEmbeddingsService,
      'generateAndSaveEmbeddings',
    )
    const sendMessageSpy = jest.spyOn(context.sqsService, 'sendMessage')

    // Triggering event handling manually for the sake of the test
    // await context.userEventsQueue.processJobs()

    expect(sendWelcomeEmailSpy).toHaveBeenCalledWith(
      savedUser.userId,
      savedUser.email,
    )
    expect(generateEmbeddingsSpy).toHaveBeenCalledWith(savedUser)
    expect(sendMessageSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        userId: savedUser.userId,
        type: EventTypeEnum.UserCreated,
      }),
    )

    // 5. User data dto should have been returned in the response (use snapshot to validate).
    expect(response.body).toMatchSnapshot()
  })
})
