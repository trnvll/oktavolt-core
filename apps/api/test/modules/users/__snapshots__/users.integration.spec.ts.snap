// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UsersController (e2e) > should create multiple users and trigger the appropriate events 1`] = `
[
  {
    "authzUserId": null,
    "context": null,
    "deletedAt": null,
    "email": "Peyton59@yahoo.com",
    "firstName": "Carley",
    "lastName": "Crist",
    "phone": "(248) 461-7600",
  },
]
`;
